src/core/iterstate.rs

```diff
@@ -418,23 +423,23 @@ mod tests {
 
         state.cost(new_cost);
 
-        assert_eq!(state.get_cost(), new_cost);
-        assert_eq!(state.get_prev_cost(), cost);
+        assert_relative_eq!(state.get_cost(), new_cost, epsilon = f64::EPSILON);
```

Comment:
My intention was to make sure that both numbers are exactly the same. I was not aware that such a comparison should be done using `.to_ne_bytes()`. Could you elaborate why this is preferable over comparing floats directly, please? I'm curious because I always thought that comparing floats directly boils down to bitwise comparison. 
However, for the sake of this test, `assert_relative_eq!` is fine too.

Reply:
Direct equality _would_ do a byte comparison on two numbers. If that's what is intended here though, it isn't clear (especially since a lot of argmin operations may choose any floating point type, even if this test is `f64`).

A common error in floating point comparisons is to compare two floats directly (e.g. `0.3f64` to `0.1f64`). However, this is almost exclusively wrong and conveys the wrong intent. Because of the way floats are represented, a computed value of `0.3` will not necessarily be "byte equal" to the literal `0.3`. 

So there's basically two ways you can compare these:

* Are they the exact same bytes?
* Are these numbers approximately equal (relative to some error margin)?

In the first case you want to compare the bytes, so the correct way to write the code would be:

```rust
assert_eq!(state.get_cost().to_ne_bytes(), new_cost.to_ne_bytes());
```

Using `to_ne_bytes` gives us two things:

1. It avoids the `clippy` lint in our tests (and I would argue you should lint your tests and examples the same way you lint your library, but maybe that's a personal preference). You probably didn't see this because you just called `cargo clippy`, which won't lint the tests or examples by default. You'll want `cargo clippy --all-features --all-targets -- -D warnings`. At the very least, `--all-targets` is needed here to get clippy to look through your tests.
2. It signals to the reader of the code that you're checking for byte-equality, not an approximation.

The second is the most important factor. If you expect a sentinel value (e.g. INFINITY, NEG_INFINITY, NAN) then byte comparisons are your best shot at that, and the code should communicate this. Case in point: I assumed these values could be approximate and would not return the exact bytes necessary, so I went for option two immediately.

**NOTE**: You could also compare `to_le_bytes` or `to_be_bytes`, but the native endian (ne) encoding is a better bet because it won't require any byte swizzling under the hood.

If you do want to check equality relative to some error epsilon, then the code I used above does what you're looking for. I am happy to change these if you think byte-equality is the proper check here. Is that the desired behaviour? 

**Another note about clippy**: Testing here, it seems that `cargo clippy --all-targets` is broken, due to missing some required features in the examples. It's possible to specify required features for each example but that makes the `Cargo.toml` somewhat noisy. I can also wrap that into this branch, but an easy way around it is just to use `--all-features` as well.

Reply:
Thanks a lot for your detailed explanation! 

> You probably didn't see this because you just called cargo clippy, which won't lint the tests or examples by default.

You are absolutely right. :facepalm: I will adapt this in the CI soon.

> A common error in floating point comparisons is to compare two floats directly (e.g. 0.3f64 to 0.1f64). However, this is almost exclusively wrong and conveys the wrong intent. Because of the way floats are represented, a computed value of 0.3 will not necessarily be "byte equal" to the literal 0.3.

I am aware of that, but in this case the numbers are not the result of a computation but instead are set manually. 
I would prefer to check for byte-equality since the numbers should not change in any way. I absolutely agree with you though that the way I have done it does not show my intention. Therefore I would tend towards the `.to_ne_bytes()` solution, because like you said, it makes it clear that byte-equality is intended.

Reply:
Excellent. Given that, I'll push a fixup for this to do byte comparison tomorrow morning as well.

Cheers, and thanks for helping get to the final solution here.

Reply:
See f783b9b09e31089651574c503a1bb298bc27bd07

Same as before, if the commit is good to go, I can autosquash and force-push before the merge.